correios/models/posting.py

```diff
@@ -407,69 +423,101 @@ def validate(cls,
+
+    @classmethod
+    def _validate_envelope(cls, width, height, length, diameter) -> list:
+        if any([width, height, length, diameter]):
+            msg = "Envelope package does not use width, height, length and diameter"
+            return [errors.PackageError(msg)]
+        return []
+
+    @classmethod
+    def _validate_box(cls, width, height, length, diameter) -> list:
+        error_list = []
+
+        if diameter:
+            msg = "Box package does not use diameter"
+            error = errors.PackageError(msg)
+            error_list.append(error)
+
+        call_args = [
+            ("width", width, MAX_WIDTH),
+            ("height", height, MAX_HEIGHT),
+            ("length", length, MAX_LENGTH),
+            ("sum of dimensions", width + height + length, MAX_SIZE)
+        ]
+        for args in call_args:
+            error = cls._validate_dimension(*args)
+            if error:
+                error_list.append(error)
+
+        return error_list
+
+    @classmethod
+    def _validate_cylinder(cls, width, height, length, diameter) -> list:
+        error_list = []
+
         if width or height:
-            raise exceptions.InvalidPackageDimensionsError(
-                "Cylinder does not use width/height: {}x{}".format(width, height)
-            )
+            msg = "Cylinder package does not use width and height"
+            error = errors.PackageError(msg)
+            error_list.append(error)
+
+        call_args = [
+            ("length", length, MAX_CYLINDER_LENGTH),
+            ("diameter", diameter, MAX_DIAMETER),
+            ("cylinder size", length + 2 * diameter, MAX_CYLINDER_SIZE),
+        ]
+        for args in call_args:
+            error = cls._validate_dimension(*args)
+            if error:
+                error_list.append(error)
 
-        cls._validate_dimension("length", length, MAX_CYLINDER_LENGTH)
-        cls._validate_dimension("diameter", diameter, MAX_DIAMETER)
-        cls._validate_dimension("cylinder size", length + 2 * diameter, MAX_CYLINDER_SIZE)
+        return error_list
 
     @classmethod
-    def _validate_dimension(cls, name, value, maximum):
-        msg = "Invalid {} (range 0~{}): {}".format(name, maximum, value)
-        if value <= 0:
-            raise exceptions.InvalidMinPackageDimensionsError(msg)
+    def _validate_dimension(cls, name, value, maximum) -> Union[errors.PackageError, None]:
+        if value <= 0 or value > maximum:
+            return errors.PackageDimensionError(name, value, 0, maximum)
 
-        if value > maximum:
-            raise exceptions.InvalidMaxPackageDimensionsError(msg)
+        return None
 
     @classmethod
-    def _validate_weight(cls, weight, service: Optional[Union[Service, str, int]]=None) -> None:
+    def _validate_weight(cls,
+                         weight,
+                         service: Optional[Union[Service, str, int]]=None) -> Union[errors.PackageError, None]:
```

Comment:
I think the interface of _validate_* methods are weird. I really don't like of methods that returns 2 types of objects because we always need an `if` statement to handle the return of this kind of method.

It's ok to return a "flag-like" value (`None` in this case) when we're coding in a language that does not implement a exception handling mechanism (like C or Go) but in languages with this mechanism it's weird and cause a lot of issues related with `AttributeError: 'NoneType' object has no attribute 'foo'` (python) or `java.lang.NullPointerException` (java).

I think we need to rethink this interface to:

1. Change `._validate_*` methods to always return a list object with errors (0 or 1 element in list) *or*...
2. Stop using the concept of `errors` in validation and use the concept of `ValidationResult` where some of these objects are "Ok" (and evaluate as a boolean `True`) and others are "Not Ok" (and evaluate as a boolean `False`), so `.validate_*()` methods can return one of these results. Then we can use a list of these results like: 

```python
if not all(error_list):
    raise PackagePackageMeasureError(error_list)
```

Reply:
Working on it too.